File: FortniteFestival.Core/Services/HttpErrorHelper.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace FortniteFestival.Core.Services
{
    public static class HttpErrorHelper
    {
        // Central error code counts (shared across service instances)
        private static readonly ConcurrentDictionary<string,int> _errorCodeCounts = new ConcurrentDictionary<string,int>(StringComparer.OrdinalIgnoreCase);

        public static (string errorCode, string errorMessage) ExtractError(string json)
        {
            if (string.IsNullOrWhiteSpace(json) || !json.TrimStart().StartsWith("{")) return (null,null);
            try
            {
                using (var doc = JsonDocument.Parse(json))
                {
                var root = doc.RootElement;
                string code = null;
                string msg = null;
                if (root.TryGetProperty("errorCode", out var ecProp) && ecProp.ValueKind == JsonValueKind.String)
                    code = ecProp.GetString();
                // Multiple patterns for message text
                if (root.TryGetProperty("errorMessage", out var emProp) && emProp.ValueKind == JsonValueKind.String)
                    msg = emProp.GetString();
                else if (root.TryGetProperty("message", out var mProp) && mProp.ValueKind == JsonValueKind.String)
                    msg = mProp.GetString();
                if(!string.IsNullOrEmpty(code)) _errorCodeCounts.AddOrUpdate(code,1,(_,v)=>v+1);
                return (code,msg);
                }
            }
            catch { return (null,null); }
        }

        public static string FormatHttpError(string op, HttpResponseMessage res, string body, string errorCode, string errorMessage)
        {
            var status = (int)res.StatusCode;
            string snippet = body==null?"<no-body>": body.Replace('\n',' ').Replace('\r',' ');
            if (snippet.Length > 180) snippet = snippet.Substring(0,180)+"...";
            var sb = new StringBuilder();
            sb.Append('[').Append(op).Append("] HTTP ").Append(status).Append(' ').Append('(').Append(res.StatusCode).Append(')');
            sb.Append(' ').Append("errorCode=").Append(errorCode??"<none>");
            if(!string.IsNullOrWhiteSpace(errorMessage)) sb.Append(' ').Append("msg=\"").Append(Truncate(errorMessage,120)).Append('"');
            sb.Append(' ').Append("bodySnippet=").Append(snippet);
            return sb.ToString();
        }

        private static string Truncate(string s, int len) => string.IsNullOrEmpty(s) ? s : (s.Length<=len? s : s.Substring(0,len)+"...");

        public static string BuildSummaryLine()
        {
            if(_errorCodeCounts.IsEmpty) return "ErrorCodeSummary: <none>";
            var parts = _errorCodeCounts.OrderByDescending(kv=>kv.Value).Select(kv=>$"{kv.Key}={kv.Value}");
            return "ErrorCodeSummary: "+ string.Join(", ", parts);
        }

        // Snapshot of counts for persistence/reporting
        public static IReadOnlyDictionary<string,int> GetErrorCountsSnapshot()
        {
            if(_errorCodeCounts.IsEmpty) return new Dictionary<string,int>();
            return _errorCodeCounts.ToDictionary(kv=>kv.Key, kv=>kv.Value, StringComparer.OrdinalIgnoreCase);
        }

        public static string ComputeCorrelationId(Exception ex)
        {
            if(ex==null) return "00000000";
            string sig = ex.GetType().FullName + "|" + ex.Message + "|" + (ex.StackTrace==null?"": new string(ex.StackTrace.Take(200).ToArray()));
            byte[] data = Encoding.UTF8.GetBytes(sig);
            using(var sha = SHA256.Create())
            {
                var hash = sha.ComputeHash(data);
                var sb = new StringBuilder();
                for(int i=0;i<hash.Length && i<4;i++) sb.Append(hash[i].ToString("X2")); // 8 hex chars
                return sb.ToString();
            }
        }
    }
}

File: FortniteFestivalRN/src/core/httpErrorHelper.ts
type ExtractedError = {errorCode?: string; errorMessage?: string};

// Lightweight in-memory counter (process-scoped) similar to the C# static dictionary.
const errorCodeCounts = new Map<string, number>();

export const __resetForTests = (): void => {
  errorCodeCounts.clear();
};

export const extractError = (json: string | null | undefined): ExtractedError => {
  if (!json) return {};
  const trimmed = json.trim();
  if (!trimmed.startsWith('{')) return {};
  try {
    const root = JSON.parse(trimmed) as Record<string, unknown>;
    const errorCode = typeof root.errorCode === 'string' ? root.errorCode : undefined;
    const errorMessage =
      typeof root.errorMessage === 'string'
        ? root.errorMessage
        : typeof root.message === 'string'
          ? root.message
          : undefined;
    if (errorCode) errorCodeCounts.set(errorCode, (errorCodeCounts.get(errorCode) ?? 0) + 1);
    return {errorCode, errorMessage};
  } catch {
    return {};
  }
};

const truncate = (s: string, len: number): string => (s.length <= len ? s : `${s.slice(0, len)}...`);

export const formatHttpError = (params: {
  op: string;
  status: number;
  statusText?: string;
  body?: string | null;
  errorCode?: string;
  errorMessage?: string;
}): string => {
  const snippetRaw = (params.body ?? '<no-body>').replace(/\r|\n/g, ' ');
  const snippet = truncate(snippetRaw, 180);
  const parts = [
    `[${params.op}] HTTP ${params.status}${params.statusText ? ` (${params.statusText})` : ''}`,
    `errorCode=${params.errorCode ?? '<none>'}`,
  ];
  if (params.errorMessage && params.errorMessage.trim().length > 0) {
    parts.push(`msg="${truncate(params.errorMessage, 120)}"`);
  }
  parts.push(`bodySnippet=${snippet}`);
  return parts.join(' ');
};

export const buildSummaryLine = (): string => {
  if (errorCodeCounts.size === 0) return 'ErrorCodeSummary: <none>';
  const parts = [...errorCodeCounts.entries()]
    .sort((a, b) => b[1] - a[1])
    .map(([k, v]) => `${k}=${v}`);
  return `ErrorCodeSummary: ${parts.join(', ')}`;
};

// Portable correlation id: 8 hex chars from an FNV-1a 32-bit hash.
export const computeCorrelationId = (err: unknown): string => {
  const input = (() => {
    if (!err) return '';
    if (err instanceof Error) return `${err.name}|${err.message}|${String(err.stack ?? '').slice(0, 200)}`;
    return String(err);
  })();

  let hash = 0x811c9dc5;
  for (let i = 0; i < input.length; i++) {
    hash ^= input.charCodeAt(i);
    hash = (hash * 0x01000193) >>> 0;
  }
  return hash.toString(16).toUpperCase().padStart(8, '0').slice(0, 8);
};
